File: fetch_product_titles_main.js
const fs = require('fs');
const WooCommerceRestApi = require("@woocommerce/woocommerce-rest-api").default;

// Load configuration from config.json
let config;
try {
    config = JSON.parse(fs.readFileSync('config.json', 'utf8'));
    console.log('✓ Loaded configuration from config.json');
    
    // Validate config structure
    if (!config.websites || !config.default_website) {
        throw new Error('Invalid config structure - missing websites or default_website');
    }
    
    const websiteConfig = config.websites[config.default_website];
    if (!websiteConfig) {
        throw new Error(`Default website '${config.default_website}' not found in config`);
    }
    
} catch (error) {
    console.error('✗ Error loading config.json:', error.message);
    process.exit(1);
}

const websiteConfig = config.websites[config.default_website];

// Initialize WooCommerce API
const api = new WooCommerceRestApi({
    url: websiteConfig.SITE_URL,
    consumerKey: websiteConfig.CONSUMER_KEY,
    consumerSecret: websiteConfig.CONSUMER_SECRET,
    version: "wc/v3"
});
console.log(`✓ Connected to WooCommerce API at ${websiteConfig.SITE_URL}`);

// Ensure data directory exists
if (!fs.existsSync('data')) {
    fs.mkdirSync('data');
}

// Create CSV write stream
const csvStream = fs.createWriteStream('data/product_titles.csv');

// Function to fetch all product titles
async function fetchAllProductTitles() {
    try {
        let page = 1;
        let totalProducts = 0;
        const perPage = 50; // Using 50 to match Python script
        console.log('\n🔄 Starting to fetch product titles...\n');

        while (true) {
            process.stdout.write(`📥 Fetching page ${page}... `);
            
            try {
                const response = await api.get("products", {
                    per_page: perPage,
                    page: page
                });

                const products = response.data;
                
                if (products.length === 0) {
                    if (page === 1) {
                        console.log('❌ No products found or error occurred.');
                    } else {
                        console.log('✓ No more products to fetch.');
                    }
                    break;
                }

                // Write titles to CSV
                products.forEach(product => {
                    // Escape quotes and handle commas in titles
                    const escapedTitle = product.name.replace(/"/g, '""');
                    csvStream.write(`"${escapedTitle}"\n`);
                });

                totalProducts += products.length;
                console.log(`✓ Added ${products.length} titles (Total: ${totalProducts})`);

                if (products.length < perPage) {
                    console.log('\n✓ Reached the last page.');
                    break;
                }

                page++;
                // Add delay to avoid hitting API rate limits
                await new Promise(resolve => setTimeout(resolve, 1000));
            } catch (error) {
                console.log(`❌ Error on page ${page}:`, error.response?.data?.message || error.message);
                break;
            }
        }

        // Close CSV file
        csvStream.end();
        console.log(`\n✅ Finished! Total products processed: ${totalProducts}`);
        console.log(`📁 Results saved to data/product_titles.csv`);

    } catch (error) {
        console.error("\n❌ Fatal Error:", error.response?.data?.message || error.message);
        // Ensure CSV is closed on error
        if (!csvStream.closed) {
            csvStream.end();
        }
    }
}

// Handle process termination
process.on('SIGINT', () => {
    console.log('\n⚠️ Process interrupted - cleaning up...');
    if (!csvStream.closed) {
        csvStream.end();
    }
    process.exit(0);
});

process.on('SIGTSTP', () => {
    console.log('\n⚠️ Process paused - cleaning up...');
    if (!csvStream.closed) {
        csvStream.end();
    }
    process.exit(0);
});

// Run the function
console.log('🚀 Starting product title import...\n');
fetchAllProductTitles();

File: fetch_product_titles_main.py
import requests
import time
import json
import csv
import os
import sys
import argparse

from dotenv import load_dotenv

# Load environment variables
load_dotenv()

def load_configuration(website_name=None):
    """Load configuration from environment variables"""
    config = {
        'CONSUMER_KEY': os.getenv('CONSUMER_KEY'),
        'CONSUMER_SECRET': os.getenv('CONSUMER_SECRET'),
        'SITE_URL': os.getenv('SITE_URL'),
        'DOMAIN': os.getenv('DOMAIN_1')
    }
    print('✓ Loaded configuration from environment variables')
    
    # Validate configuration
    required_keys = ['CONSUMER_KEY', 'CONSUMER_SECRET', 'SITE_URL']
    missing_keys = [key for key in required_keys if not config.get(key)]
    if missing_keys:
        print(f"❌ Error: Missing required configuration: {', '.join(missing_keys)}")
        exit(1)
    
    return config

CSV_FILE = "data/product_titles.csv"
PAGE_PROPERTY_FILE = "data/product_titles_page.txt"  # File to store the current page number


def get_current_page():
    """Retrieve the current page from the property file."""
    try:
        with open(PAGE_PROPERTY_FILE, "r") as file:
            return int(file.read().strip())
    except FileNotFoundError:
        return 1
    except ValueError:
        print("⚠️ Warning: Invalid page number in current_page.txt. Starting from page 1.")
        return 1


def save_current_page(page):
    """Save the current page number to the property file."""
    try:
        with open(PAGE_PROPERTY_FILE, "w") as file:
            file.write(str(page))
    except IOError as e:
        print(f"⚠️ Warning: Could not save current page: {e}")


def fetch_titles(page, config):
    """Fetch product titles from WooCommerce API."""
    api_url = f"{config['SITE_URL']}/wp-json/wc/v3/products"
    params = {
        'page': page,
        'per_page': 50,
        'consumer_key': config['CONSUMER_KEY'],
        'consumer_secret': config['CONSUMER_SECRET']
    }
    
    try:
        response = requests.get(api_url, params=params)
        response.raise_for_status()
        products = response.json()
        return [product["name"] for product in products], len(products) == 50
    except requests.RequestException as e:
        print(f"❌ Error fetching data from API: {e}")
        return [], False
    except (KeyError, json.JSONDecodeError) as e:
        print(f"❌ Error processing API response: {e}")
        return [], False


def write_titles_to_csv(titles):
    """Write titles to a CSV file."""
    try:
        # Ensure data directory exists
        os.makedirs(os.path.dirname(CSV_FILE), exist_ok=True)
        # Open in append mode to add new titles
        mode = 'a' if get_current_page() > 1 else 'w'
        with open(CSV_FILE, mode, newline='', encoding='utf-8') as f:
            writer = csv.writer(f, quoting=csv.QUOTE_ALL)
            # Write header if it's a new file
            if mode == 'w':
                writer.writerow(['Product Title'])
            # Write titles
            for title in titles:
                writer.writerow([title])
        print(f"✓ Successfully wrote {len(titles)} titles to {CSV_FILE}")
    except Exception as e:
        print(f"❌ Error writing to CSV file: {e}")


def fetch_woocommerce_product_titles(config, website_name="default"):
    """Main function to fetch product titles and save them to CSV."""
    print(f"\n🔄 Starting to fetch product titles from {config['SITE_URL']}")
    if website_name != "default":
        print(f"📊 Website: {website_name}")
    
    current_page = get_current_page()
    total_products = 0

    while True:
        print(f"📥 Fetching page {current_page}...", end=' ', flush=True)
        titles, has_more = fetch_titles(current_page, config)
        
        if not titles:
            if current_page == 1:
                print("❌ No products found or error occurred.")
                break
            print("✓ No more products to fetch.")
            break

        total_products += len(titles)
        write_titles_to_csv(titles)
        save_current_page(current_page)

        if not has_more:
            print("✓ Reached the last page.")
            break

        current_page += 1
        time.sleep(1)  # Add delay to avoid hitting API rate limits

    print(f"\n✅ Finished! Total products processed: {total_products}")
    print(f"📁 Results saved to {CSV_FILE}")


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description='Fetch WooCommerce product titles with multi-website support')
    parser.add_argument('--website', type=str, help='Website name (from config.json)')
    
    args = parser.parse_args()
    
    print('🚀 Starting product title import...\n')
    try:
        config = load_configuration(args.website)
        fetch_woocommerce_product_titles(config, args.website or "default")
    except KeyboardInterrupt:
        print("\n⚠️ Process interrupted by user. Progress has been saved.")
    except Exception as e:
        print(f"❌ An unexpected error occurred: {e}")
